akka-stream/src/main/scala/akka/stream/impl/io/IOSinks.scala

```diff
@@ -66,3 +68,72 @@ private[akka] final class OutputStreamSink(createOutput: () ⇒ OutputStream, va
   override def withAttributes(attr: Attributes): Module =
     new OutputStreamSink(createOutput, attr, amendShape(attr))
 }
+
+/**
+ * INTERNAL API
+ * Creates simple synchronous (Java 6 compatible) Sink which provide all incoming element on read request via
+ * InputStream.
+ */
+private[akka] final class InputStreamSink(val timeout: FiniteDuration, val attributes: Attributes, shape: SinkShape[ByteString])
+  extends SinkModule[ByteString, (InputStream, Future[Long])](shape) {
+
+  override def create(context: MaterializationContext) = {
+    val mat = ActorMaterializer.downcast(context.materializer)
+    val settings = mat.effectiveSettings(context.effectiveAttributes)
+    val bytesWrittenPromise = Promise[Long]()
+
+    val props = InputStreamSubscriber.props(bytesWrittenPromise, settings.maxInputBufferSize)
+
+    val ref = mat.actorOf(context, props)
+    (akka.stream.actor.ActorSubscriber[ByteString](ref),
+      (new InputStreamAdapter(ref, timeout).asInstanceOf[InputStream], bytesWrittenPromise.future))
+  }
+
+  override protected def newInstance(shape: SinkShape[ByteString]): SinkModule[ByteString, (InputStream, Future[Long])] =
+
+    new InputStreamSink(timeout, attributes, shape)
+
+  override def withAttributes(attr: Attributes): Module =
+    new InputStreamSink(timeout, attr, amendShape(attr))
+}
+
+/**
+ * INTERNAL API
+ * InputStreamAdapter that interacts with InputStreamSubscriber
+ */
+sealed private class InputStreamAdapter(subscriber: ActorRef, val timeout: FiniteDuration) extends InputStream {
+
+  import akka.pattern.ask
+
+  @scala.throws(classOf[IOException])
+  override def read(): Int = {
+    val a = new Array[Byte](1)
+    read(a)
+    a(0)
+  }
+
+  @scala.throws(classOf[IOException])
+  override def read(a: Array[Byte]): Int = read(a, 0, a.length)
+
+  @scala.throws(classOf[IOException])
+  override def read(a: Array[Byte], begin: Int, length: Int): Int = {
+    val bytes = sendMessage(InputStreamSubscriber.Read(length)).toArray
+    System.arraycopy(bytes, 0, a, begin, bytes.length)
+    bytes.length
+  }
+
+  @scala.throws(classOf[IOException])
+  override def close(): Unit = subscriber.tell(InputStreamSubscriber.Close(), ActorRef.noSender)
```

Comment:
maybe also via `sendMessage`? What if close fails, I guess I'd like to know when using this stream